### DISCUSSION TITLE
Docker image with requirements for "Strategies for Pre-training Graph Neural Networks"

### DISCUSSION THREAD

**Discussion by mdanb** [pyg-team/pytorch_geometric#2322] [Q&A] [answered]
I am trying to run the [code](https://github.com/snap-stanford/pretrain-gnns) from the [paper](https://arxiv.org/abs/1905.12265) "Strategies for Pre-training Graph Neural Networks". The code requires a really old version of Pytorch: Pytorch 1.0.1
I'm trying to run the code from a Docker container. So I first pulled the [pytorch/pytorch:1.0.1-cuda10.0-cudnn7-runtime](https://hub.docker.com/layers/pytorch/pytorch/1.0.1-cuda10.0-cudnn7-runtime/images/sha256-e19f3b87dbf3db7a2492d64eb4520dbdc7ed368e33203ea48c8393c549d49501?context=explore) docker image. I then installed all requirements, including `torch-geometric` and `torch-scatter` etc... and committed to create a new image with the necessary requirements
I then tried running the code in the container, but I get the error:
`ModuleNotFoundError: No module named 'torch_scatter.scatter_cuda'`
Was wondering how I can solve this?

**Comment by rusty1s** [comment, marked as answer]
Here is an [old Dockerfile](https://github.com/rusty1s/pytorch_geometric/blob/1.1.0/docker/Dockerfile) which might be helpful to you.

Nonetheless, I don't think there were any breaking changes from PyTorch 1.0.1 and PyTorch Geometric 1.0.3. It should be pretty straightforward to update the code to support later versions of them.

**Reply by mdanb** [reply]
Actually, I have a few questions about that since that's what I tried to do initially. First, do you mean Pytorch Geometric 1.3.0 instead of 1.0.3? Because the authors also mention 1.0.3 instead of 1.3.0 which confused me because I couldn't find 1.0.3
So I then decided just to search the 1.3.0 documentation. I was looking for a method that works on `Data` objects called `cat_dim` since the authors' code includes the line:
`batch[key] = torch.cat(batch[key], dim=data_list[0].cat_dim(key, batch[key][0]))`
and when I would run it I would get:
`AttributeError: 'Data' object has no attribute 'cat_dim'`
I couldn't find a `cat_dim`, but I found a `__cat_dim__` instead (are they the same?). I then just replaced the former by the latter, which did make the error go away, but then I got another error saying:
```ValueError: MessagePassing.propagate only supports torch.LongTensor of shape [2, num_messages] or torch_sparse.SparseTensor for argument edge_index.```

Now, right after the line above they have:
```
batch.batch = torch.cat(batch.batch, dim=-1)
 return batch.contiguous()
```
and I added the line:
`
batch.edge_index = batch.edge_index.to_sparse()
`
between them
But then I got 
`RuntimeError: sparse tensors do not have is_contiguous`

I then finally tried just doing `return batch` instead of `return batch.contiguous()`
Then, the line `edge_index = add_self_loops(edge_index, num_nodes = x.size(0))` in the code was causing issues:
```File "/home/mdanb/.local/lib/python3.8/site-packages/torch_geometric/utils/loop.py", line 92, in add_self_loops
    edge_index = torch.cat([edge_index, loop_index], dim=1)
RuntimeError: Concatenating sparse tensors, but a dense tensor was found at position 1.
```

I stopped trying at that point haha because I figured the problem was maybe in the initial change from `cat_dim` to `__cat_dim__` but I don't know I'm not sure

**Reply by rusty1s** [reply]
Indeed. `cat_dim` got renamed to `__cat_dim__`.

The second error relates to a change in the `propagate` API. `self.aggr` is now an argument passed to the constructor, i.e., `super(GINConv, self).__init__(aggr=aggr)` and the `propagate` call should look like:
```python
return self.propagate(edge_index, x=x, edge_attr=edge_embeddings)
```
